exporter/clickhouseexporter/exporter_metrics.go

```diff
@@ -0,0 +1,110 @@
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/clickhouseexporter/internal"
+)
+
+type metricsExporter struct {
+	client *sql.DB
+
+	logger *zap.Logger
+	cfg    *Config
+}
+
+func newMetricsExporter(logger *zap.Logger, cfg *Config) (*metricsExporter, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+	if err := createDatabase(cfg); err != nil {
+		return nil, err
+	}
+	client, err := newClickhouseClient(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = internal.CreateMetricsTable(cfg.MetricsTableName, cfg.TTLDays, client); err != nil {
+		return nil, err
+	}
+
+	return &metricsExporter{
+		client: client,
+		logger: logger,
+		cfg:    cfg,
+	}, nil
+}
+
+// Shutdown will shutdown the exporter.
+func (e *metricsExporter) shutdown(ctx context.Context) error {
+	if e.client != nil {
+		return e.client.Close()
+	}
+	return nil
+}
+
+func (e *metricsExporter) pushMetricsData(ctx context.Context, md pmetric.Metrics) error {
+	err := doWithTx(ctx, e.client, func(tx *sql.Tx) error {
+		metricsMap := internal.CreateMetricsModel(e.cfg.MetricsTableName)
+		for i := 0; i < md.ResourceMetrics().Len(); i++ {
+			metaData := internal.MetricsMetaData{}
+			metrics := md.ResourceMetrics().At(i)
+			res := metrics.Resource()
+			metaData.ResAttr = attributesToMap(res.Attributes())
+			metaData.ResURL = metrics.SchemaUrl()
+			for j := 0; j < metrics.ScopeMetrics().Len(); j++ {
+				rs := metrics.ScopeMetrics().At(j).Metrics()
+				metaData.ScopeURL = metrics.ScopeMetrics().At(j).SchemaUrl()
+				metaData.ScopeInstr = metrics.ScopeMetrics().At(j).Scope()
+				for k := 0; k < rs.Len(); k++ {
+					r := rs.At(k)
+					switch r.Type() {
+					case pmetric.MetricTypeGauge:
+						metricsMap[pmetric.MetricTypeGauge].Add(r.Gauge(), r.Name(), r.Description(), r.Unit())
+					case pmetric.MetricTypeSum:
+						metricsMap[pmetric.MetricTypeSum].Add(r.Sum(), r.Name(), r.Description(), r.Unit())
+					case pmetric.MetricTypeHistogram:
+						metricsMap[pmetric.MetricTypeHistogram].Add(r.Histogram(), r.Name(), r.Description(), r.Unit())
+					case pmetric.MetricTypeExponentialHistogram:
+						metricsMap[pmetric.MetricTypeExponentialHistogram].Add(r.ExponentialHistogram(), r.Name(), r.Description(), r.Unit())
+					case pmetric.MetricTypeSummary:
+						metricsMap[pmetric.MetricTypeSummary].Add(r.Summary(), r.Name(), r.Description(), r.Unit())
+					default:
+						return fmt.Errorf("unsupported metrics type")
+					}
+				}
+				internal.InjectMetaData(metricsMap, &metaData)
+			}
+		}
+
+		// batch insert https://clickhouse.com/docs/en/about-us/performance/#performance-when-inserting-data
+		if err := internal.InsertMetrics(ctx, tx, metricsMap, e.logger); err != nil {
+			return exporterhelper.NewThrottleRetry(err, time.Duration(0)*time.Second)
```

Comment:
`time.Duration(0)*time.Second` ?

Reply:
actully I don't know show to use this help function, I follow this:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/dec0e4791fd6216442d06cb8e743c1c82222ec5c/exporter/logzioexporter/exporter.go#L253-L264

Reply:
@atoulme can you give me a example how to use it?

Reply:
I haven't used it before either. I would grep the repository for examples of use. You're not really following this guidance since you're not using the `Retry-After` header here, just going to 0 directly.

Reply:
I take a while to find the examples of `exporterhelper.NewThrottleRetry`, but I found that it's used for HTTP or gRPC scene, When the `tx.ExecContext` returns an error, I don't think it will return any header messages.  Can I return the error like [here does](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/clickhouseexporter/exporter_logs.go#L118)? I want to put the retry feature in the future and find a general way to deal with logs, traces, and metrics.